pkg/cloudprovider/providers/vsphere/vsphere.go

```diff
@@ -238,228 +161,80 @@ func init() {
-// * sysfs entry
-func getVMName(client *govmomi.Client, cfg *VSphereConfig) (string, error) {
-	var vmUUID string
-	var err error
-
-	if cfg.Global.VMUUID != "" {
-		vmUUID = cfg.Global.VMUUID
-	} else {
-		// This needs root privileges on the host, and will fail otherwise.
-		vmUUID, err = getvmUUID()
-		if err != nil {
-			return "", err
-		}
-		cfg.Global.VMUUID = vmUUID
-	}
-
-	if vmUUID == "" {
-		return "", fmt.Errorf("unable to determine machine ID from cloud configuration or sysfs")
-	}
-
-	// Create context
-	ctx, cancel := context.WithCancel(context.Background())
-	defer cancel()
-
-	// Create a new finder
-	f := find.NewFinder(client.Client, true)
-
-	// Fetch and set data center
-	dc, err := f.Datacenter(ctx, cfg.Global.Datacenter)
-	if err != nil {
-		return "", err
-	}
-	f.SetDatacenter(dc)
-
-	s := object.NewSearchIndex(client.Client)
-
-	svm, err := s.FindByUuid(ctx, dc, strings.ToLower(strings.TrimSpace(vmUUID)), true, nil)
-	if err != nil {
-		return "", err
-	}
-
-	if svm == nil {
-		return "", fmt.Errorf("unable to find machine reference by UUID")
-	}
-
-	var vm mo.VirtualMachine
-	err = s.Properties(ctx, svm.Reference(), []string{"name"}, &vm)
-	if err != nil {
-		return "", err
-	}
-
-	return vm.Name, nil
-}
-
 func newVSphere(cfg VSphereConfig) (*VSphere, error) {
+	var err error
 	if cfg.Disk.SCSIControllerType == "" {
-		cfg.Disk.SCSIControllerType = PVSCSIControllerType
-	} else if !checkControllerSupported(cfg.Disk.SCSIControllerType) {
+		cfg.Disk.SCSIControllerType = vclib.PVSCSIControllerType
+	} else if !vclib.CheckControllerSupported(cfg.Disk.SCSIControllerType) {
 		glog.Errorf("%v is not a supported SCSI Controller type. Please configure 'lsilogic-sas' OR 'pvscsi'", cfg.Disk.SCSIControllerType)
 		return nil, errors.New("Controller type not supported. Please configure 'lsilogic-sas' OR 'pvscsi'")
 	}
 	if cfg.Global.WorkingDir != "" {
-		cfg.Global.WorkingDir = path.Clean(cfg.Global.WorkingDir) + "/"
+		cfg.Global.WorkingDir = path.Clean(cfg.Global.WorkingDir)
 	}
 	if cfg.Global.RoundTripperCount == 0 {
 		cfg.Global.RoundTripperCount = RoundTripperDefaultCount
 	}
 	if cfg.Global.VCenterPort != "" {
 		glog.Warningf("port is a deprecated field in vsphere.conf and will be removed in future release.")
 	}
-
-	var c *govmomi.Client
-	var id string
+	if cfg.Global.VMUUID == "" {
+		// This needs root privileges on the host, and will fail otherwise.
+		cfg.Global.VMUUID, err = getvmUUID()
+		if err != nil {
+			glog.Errorf("Failed to get VM UUID. err: %+v", err)
+			return nil, err
+		}
+	}
+	vSphereConn := vclib.VSphereConnection{
+		Username:          cfg.Global.User,
+		Password:          cfg.Global.Password,
+		Hostname:          cfg.Global.VCenterIP,
+		Insecure:          cfg.Global.InsecureFlag,
+		RoundTripperCount: cfg.Global.RoundTripperCount,
+	}
+	var instanceID string
+	// Create context
+	ctx, cancel := context.WithCancel(context.TODO())
+	defer cancel()
+	err = vSphereConn.Connect(ctx)
```

Comment:
I fixed it in latest commit. Now we don't connect to VC on worker nodes if we have vm-name available in vsphere.conf file.